index.js

```diff
@@ -466,7 +468,7 @@ const createFocusTrap = function (elements, userOptions) {
       const containerGroup =
         containerIndex >= 0 ? state.containerGroups[containerIndex] : undefined;
 
-      if (containerIndex < 0) {
+      if (!containerGroup) {
```

Comment:
The rest of the changes below this line are too much for me to grok. Feels like there's a lot of carefully-honed logic in there that we're deconstructing and changing around. Can you explain the changes in behavior happening in here? You're even using the modulus operator?! 😉 I'm also concerned about some key comments going away. I rely on comments to remember what was going on in some edge case and why we did something because I'm not in this on a daily basis. Would be good to retain those that are still relevant, or replace them with updated ones.

Reply:
> The rest of the changes below this line are too much for me to grok.

Let's break it down into manageable pieces.
We have the block of this `if (!containerGroup)` condition you mentioned, which I din't touch, and then two `else` blocks: one for when the user is tabbing forward and one for when they are tabbing backward. The two blocks are quite symmetric, the logic is the same, so let's consider just the block for the case in which they are tabbing forward:

https://github.com/focus-trap/focus-trap/blob/f198c2bcbcb9b09e08b7bb3a75288d0163983880/index.js#L501

This is the whole block of code:
```js
if (
  target === containerGroup.container ||
  containerGroup.focusableNodes.indexOf(target) >=
    containerGroup.focusableNodes.indexOf(
      containerGroup.lastTabbableNode
    )
) {
  // YES: then tab should go to the first tabbable node in the next
  //  group (and wrap around to the first tabbable node of the FIRST
  //  group if it's the last tabbable node of the LAST group)
  const destinationGroupIndex =
    (containerIndex + 1) % state.containerGroups.length;

  const destinationGroup = state.tabbableGroups[destinationGroupIndex];

  destinationNode =
    getTabIndex(target) >= 0
      ? destinationGroup.firstTabbableNode
      : destinationGroup.firstDomTabbableNode;
} else if (!isTabEvent(event)) {
  // user must have customized the nav keys so we have to move focus manually _within_
  //  the active group: do this based on the order determined by tabbable()
  destinationNode = containerGroup.nextTabbableNode(target);
}
```

The condition:
```js
if (
  target === containerGroup.container ||
  containerGroup.focusableNodes.indexOf(target) >=
    containerGroup.focusableNodes.indexOf(
      containerGroup.lastTabbableNode
    )
```

is assessing whether `target` is itself the `container` containing `target`; or the `lastTabbableNode` in the `container`; or a focusable node within the `container` that follows `lastTabbableNode` in document order.
If `target` is one of these elements, focus-trap can't trust the browser, it has to intervene and programmatically pass the focus to an element in the next container.
That's how we get the index of the next container:

```js
const destinationGroupIndex =
  (containerIndex + 1) % state.containerGroups.length;
```
> You're even using the modulus operator?!  

It has been waiting its whole life for this moment :laughing: it's the perfect use case for it.

So now that we have the next container, what's the right element we need within it?
As we discussed in the looong issue we are trying to close, if `target` has a negative tab index, the focus can't simply go to the `firstTabbableNode` of the container, because that takes into account tab indexes (meaning it could be a positive tab index that is the last element in document order within the container) while browsers just pick the closest tabbable element in document order, without looking at tab index values. So the element to focus is:

```js
destinationNode =
  getTabIndex(target) >= 0
    ? destinationGroup.firstTabbableNode
    : destinationGroup.firstDomTabbableNode;
```

The last couple of lines in the block (that `else if`) were already there, and I didn't touch them.
Though I did touch the body of `nextTabbableNode`, which is being called there.

Reply:
Actually I inherited the condition `target === containerGroup.container` from here:

https://github.com/focus-trap/focus-trap/blob/f22e3611a6eebbf1245f34e0a37a86250a5875fc/index.js#L533

But I notice only now that it's wrong. It is necessary only when tabbing backward, because the container itself would be like the very first focusable element in the container. But when tabbing forward, we don't care about it, because we only look at the last focusable elements.

I'm going to fix it.

Reply:
Removing `target === containerGroup.container` from the case in which they are tabbing forward makes a test fail.
I'm not sure why, an hypothesis is that in that case `findNextNavNode` is not being called on `keydown` but rather on `focusin`:

https://github.com/focus-trap/focus-trap/blob/f198c2bcbcb9b09e08b7bb3a75288d0163983880/index.js#L441

I haven't yet considered the implications of this case, and right now I really need to get some sleep.
Will look into it as soon as I can.

Reply:
Not sure. That's why I'm so hesitant to make changes in there... `findNextNavNode()` is being called on `focusin` and `keydown`, but in `focusin` only in the case where you're tabbing across containers. It's always called on `keydown`.

I'm sure that clause is there for a reason.

Reply:
Alright, I haven't even looked at where `findNextNavNode` is called, I just trust that it's better to minimise the amount of changes.

But [replying to](https://github.com/focus-trap/focus-trap/pull/987#pullrequestreview-1503490260):

>Your PR is trying to fix all cases where it's technically possible to tab away from a negative tabindex node. Probably should not be lumped into my PR. 

I think that at least the changes in https://github.com/focus-trap/focus-trap/pull/987/commits/b5a5f191a02f30e20026bef454076b9942e08a2e should be introduced:

```js
destinationNode =
  getTabIndex(target) >= 0
    ? destinationGroup.firstTabbableNode
    : destinationGroup.firstDomTabbableNode;
```

because if positive tab indexes are not supported, `firstDomTabbableNode` coincides with `firstTabbableNode` , the distinction between the two should be something to be taken care of at the moment of introducing support for positive tab indexes.

Reply:
Reverted changes in to `findNextNavNode` here: 20af8f2.
(Except for the changes [discussed above](https://github.com/focus-trap/focus-trap/pull/987#discussion_r1245715986): https://github.com/focus-trap/focus-trap/commit/b5a5f191a02f30e20026bef454076b9942e08a2e)

Reply:
Just realised that the https://github.com/focus-trap/focus-trap/commit/b5a5f191a02f30e20026bef454076b9942e08a2e won't do anything without the rest of of the changes, because now the destination node is set only [`if (startOfGroupIndex >= 0)`](https://github.com/focus-trap/focus-trap/blob/20af8f2bcedadf5256ca3da8d01e55624cebc278/index.js#L506C9-L506C37) which doesn't take into account focusable, and not tabbable, nodes.

I could fix just that, BUT... I think we can get rid of `focusin` altogether and reintroduce all the changes I was proposing.

`focusin` was needed before because the library couldn't handle certain edge cases, such as positive tab indexes, and had to deal with the focus unexpectedly escape the trap.
But once there's support for positive tab indexes (only for single container traps) there's no way I can think of to have the focus unexpectedly escape the trap (assuming the user doesn't create a multi-container trap with positive tab indexes).

It should be noted, though, that removing `focusin` would change the behaviour of the library in regard to multi-container traps with positive tab indexes, since in that case the focus could escape and there would be no `focusin` to bring it back.

You may want to leave `findNextNavNode` as it is and create a new handler just for `keydown`, then always use the new handler unless the user has a multi-container trap with positive tab indexes, in that case fallback to use the legacy logic that doesn't properly handle positive tab indexes, but still doesn't let the focus escape.

But in my humble opinion, it's unnecessary to keep support for a case that is not supported anymore (positive-tab indexes in multi-container traps).
It would be acceptable to just document that the behaviour for that edge case changed since version x.x.x

Shall we remove `focusin`?

Reply:
Sorry, even though there's no way for the focus to escape the trap via keyboard navigation, there's still the possibility that focus doesn't enter the trap to begin with.

After the user focuses an element that doesn't belong to any container and tabs away from it, the focus may be given to an element outside of the trap.
Handling this case on `keydown` requires comparing the document position of `event.target` against the containers, which is not trivial because of the shadow DOM.

Reply:
Sorry, even though there's no way for the focus to escape the trap via keyboard navigation, there's still the possibility that the focus doesn't enter the trap to begin with.

After the user focuses an element that doesn't belong to any container and tabs away from it, the focus may be given to an element outside of the trap.

Handling this case on properly requires comparing the document position of `event.target` against the containers, which is not trivial because of the shadow DOM.

And while it could be handled improperly on `keydown` rather than on `focusin`, I guess that `focusin` is the preferred time because it doesn't prevent the browser from doing the right thing when possible.

Reply:
I greatly appreciate your deep thinking here, but no, I do not support the idea of removing `focusin`. I'm 99.99999% sure that focus does not only escape due to positive tab indexes. The library has been dealing with focus escape for a very long time and I can just imagine all of what would potentially hit the fan if that handler wasn't there keeping an eye on things. 😅 

> And while the case could be handled improperly on `keydown` rather than on `focusin`, I guess that `focusin` is the preferred time because it doesn't prevent the browser from doing the right thing when possible.

I assume you meant to write "properly" instead of "improperly" here, but the second half of your sentence is the key: In as much as is possible, get out of the way and let the browser handle the focus. 🙂

Reply:
@stefcameron I actually meant to say "improperly".
If `target` doesn't belong to any container, the case can't be handled properly (because of the shadow DOM).
So it is handled improperly on `focusin`, which is better than handling it improperly on `keydown` :laughing:

Reply:
Got it. 😉